File: server/php/src/Parallelbug.php
<?php

Class Parallelbug {
	public static $dir = NULL; // No trailing slash
	public static $settings = array(
		//'filename' => uniqid(),
		'format' => 'json',
		'meta' => true,
		'meta_key' => '__meta__',
		'date_format' => 'Y-m-d H:i:s',
		'append' => false,
		'data_key' => 'dump',
		'append_group_key' => '__append_group__'
	);

	public static function dump($data, $opts) {
		// Do not accept dump if $dir is unset
		if(self::$dir === NULL) return false;

		// Default params
		$settings = self::$settings;
		$settings['filename'] = uniqid();

		$opts = array_merge($settings, $opts);

		// Construct filepath
		$filename = self::$dir . DIRECTORY_SEPARATOR . $opts['filename'];

		if($opts['meta'] === true) {
			$now = time();

			// Generate structure with meta
			$data = array(
				$opts['meta_key'] => array(
					'length' => count($data),
					'type' => gettype($data),
					'timestamp' => $now,
					'timestamp_pretty' => date($opts['date_format'], $now)
				),

				$opts['data_key'] => $data
			);
		}

		if($opts['format'] === 'json') {
			// Ensure extension
			$filename .= '.json';

			if($opts['append'] === true) {
				// Grab old dump - This is so that we can properly append the new data
				$old_data = file_get_contents($filename);
				if($old_data === false) return false;

				$_old_data = json_decode($old_data, true);

				if(in_array($opts['append_group_key'], $_old_data)) {
					// The old data is already in an append group, so just push new data
					$tmp = $_old_data;
					$tmp[] = $data;

					$data = $tmp;
					unset($tmp);
				} else {
					// The old data is not in an append group, so create structure
					$tmp = array(
						$opts['append_group_key'],
						$_old_data,
						$data
					);
					$data = $tmp;
					unset($tmp);
				}
			}

			$_data = json_encode($data);
		}

		if(!isset($_data)) return false;
		// Place dump in file. The ternary enforces accurate boolean return. Do not use !! as this will give false positives.
		return file_put_contents($filename, $_data) === false ? false : true;
	}
}

File: server/python/3.x/src/Parallelbug.py
import uuid
import os
import time
import datetime
import json

# Python 3

class Parallelbug:
	dir = None
	settings = {
		"format": "json",
		"meta": True,
		"meta_key": "__meta__",
		"date_format": "%Y-%m-%d %H:%M:%S",
		"append": False,
		"data_key": "dump",
		"append_group_key": "__append_group__"
	}

	@staticmethod
	def dump(data, opts):
		if(Parallelbug.dir is None):
			return False

		#Default params
		settings = dict(Parallelbug.settings)
		settings["filename"] = str(uuid.uuid4())

		opts = dict(list(settings.items()) + list(opts.items()))

		filename = os.path.join(Parallelbug.dir, opts["filename"])

		if(opts["meta"] == True):
			now = time.time()
			if(type(data) is dict or type(data) is list or type(data) is str):
				length = len(data)
			else:
				length = None

			data = {
				opts["meta_key"]: {
					"length": length,
					"type": type(data).__name__,
					"timestamp": now,
					"timestamp_pretty": datetime.datetime.fromtimestamp(now).strftime(opts["date_format"])
				},

				opts["data_key"]: data
			}

		if(opts["format"] == "json"):
			filename += ".json"

			if(opts["append"] == True):
				with open(filename) as infile:
				    old_data = json.load(infile)

				if(type(old_data) is list and opts["append_group_key"] in old_data):
					tmp = old_data
					tmp.append(data)
					data = tmp
				else:
					tmp = [
						opts["append_group_key"],
						old_data,
						data
					]
					data = tmp

			with open(filename, 'w') as outfile:
  				json.dump(data, outfile)